### PR TITLE
Do not create jobs until pg inqueue

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Fixes: #621

### CODE REVIEW

File: pkg/controllers/job/job_controller_actions.go [hunk 1]
```diff
@@ -185,9 +189,46 @@ func (cc *Controller) syncJob(jobInfo *apis.JobInfo, updateStatus state.UpdateSt
...
+		if e := cc.cache.Update(newJob); e != nil {
+			klog.Errorf("SyncJob - Failed to update Job %v/%v in cache:  %v",
+				newJob.Namespace, newJob.Name, e)
+			return e
+		}
+		return nil
+	}
+
+	// Skip job task sync if it is pending
+	if job.Status.State.Phase == batch.Pending {
```
Comment (k82cn) [inline pkg/controllers/job/job_controller_actions.go]: If return, when do we check it again?

Comment (hzxuzhonghu) [reply pkg/controllers/job/job_controller_actions.go]: when receives pg update event

**Comment by hzxuzhonghu** [comment]
@k82cn Please let me know if this make sense to you, so i can decide if i should work on fixing ci. Otherwise, i can close it.

**Comment by hzxuzhonghu** [comment]
I am pretty sure the e2e is flaky

File: pkg/controllers/job/job_controller_actions.go [hunk 2]
```diff
@@ -185,9 +189,38 @@ func (cc *Controller) syncJob(jobInfo *apis.JobInfo, updateStatus state.UpdateSt
 		return nil
 	}
 
-	var err error
-	if job, err = cc.createJob(job); err != nil {
-		return err
+	// Skip job initiation if job is already accepted
+	if job.Status.State.Phase == "" || job.Status.State.Phase == batch.Pending || job.Status.State.Phase == batch.Restarting {
```
Comment (k82cn) [inline pkg/controllers/job/job_controller_actions.go]: add `isInitialized` func for this check.

File: pkg/controllers/job/job_controller_actions.go [hunk 3]
```diff
@@ -185,9 +189,38 @@ func (cc *Controller) syncJob(jobInfo *apis.JobInfo, updateStatus state.UpdateSt
...
+		var err error
+		if job, err = cc.initiateJob(job); err != nil {
+			return err
+		}
+	}
+	var syncTask bool
+	if pg, _ := cc.pgLister.PodGroups(job.Namespace).Get(job.Name); pg != nil {
```
Comment (k82cn) [inline pkg/controllers/job/job_controller_actions.go]: can we check this in handler?

**Comment by k82cn** [comment]
LGTM overall :)

/approve

File: pkg/controllers/job/job_controller_actions.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/controllers/job/job_controller_actions.go]: done

File: pkg/controllers/job/job_controller_actions.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/controllers/job/job_controller_actions.go]: I am not sure i am understanding what you mean. My thought is to split the SyncJob function: the first phase is initialization and the second phase is sync.

**Pull request merged**